Signal/src/ViewControllers/ConversationView/ConversationViewController.m

```diff
@@ -2124,6 +2130,206 @@ - (void)didTapFailedOutgoingMessage:(TSOutgoingMessage *)message
     [self handleUnsentMessageTap:message];
 }
 
+- (void)didTapQuotedMessage:(ConversationViewItem *)viewItem quotedMessage:(TSQuotedMessage *)quotedMessage
+{
+    OWSAssertIsOnMainThread();
+    OWSAssert(viewItem);
+    OWSAssert(quotedMessage);
+    OWSAssert(quotedMessage.timestamp > 0);
+    OWSAssert(quotedMessage.authorId.length > 0);
+
+    // We try to range of items within the current thread's interactions
+    // that includs the "quoted interaction".
+    // NOTE: Since the range _IS NOT_ filtered by author,
+    // and timestamp collisions are possible, it's possible
+    // for:
+    //
+    // * The range to include more than the "quoted interaction".
+    // * The range to be non-empty but NOT include the "quoted interaction",
+    //   although this would be a bug.
+    NSUInteger threadInteractionCount = 0;
+    NSRange itemRange = [self findRangeOfQuotedMessage:quotedMessage threadInteractionCount:&threadInteractionCount];
```

Comment:
I'm not sure why we're finding the "range" of a single message. I think what we want is it's index, which we can get in constant time rather than using the ln(n) binary search.

1. Use the "idx" Secondary Index (`TSInteraction interactionsWithTimestamp:`) to go from `quotedMessage.timestamp` -> database key (`interaction.uniqueId`)

2. Use that key to get the view index via:
```
/**
 * Fetches both the group and the index within the group for the given {collection, key}.
 *
 * Returns YES if the {collection, key} is included in the view.
 * Otherwise returns NO, and sets the parameters to nil & zero.
**/
- (BOOL)getGroup:(NSString * _Nullable * _Nullable)groupPtr
           index:(nullable NSUInteger *)indexPtr
          forKey:(NSString *)key
    inCollection:(nullable NSString *)collection;
```

Would that work?

Reply:
Good idea, done.